### PR TITLE
Support encoding/decoding of parameterized records

### PR DESCRIPTION
**Pull request by jyemin**
JAVA-4740

### CODE REVIEW

File: bson-record-codec/src/main/org/bson/codecs/record/RecordCodec.java [hunk 1]
```diff
@@ -218,16 +254,47 @@ private static <T extends Annotation> void validateAnnotationOnlyOnField(final R
 
     RecordCodec(final Class<T> clazz, final CodecRegistry codecRegistry) {
         this.clazz = notNull("class", clazz);
+        if (clazz.getTypeParameters().length > 0) {
```
Comment (jyemin) [inline bson-record-codec/src/main/org/bson/codecs/record/RecordCodec.java:257]: In this state, the codec is unusable unless it is first parameterized.  That effectively means that we don't support encoding records where the outermost record has type parameters.  This is the same restriction as we have for POJOs.

File: bson-record-codec/src/main/org/bson/codecs/record/RecordCodec.java [hunk 2]
```diff
@@ -109,6 +115,36 @@ private static Codec<?> computeCodec(final RecordComponent component, final Code
...
+        }
+
+        private static Type resolveType(final Type type, final List<Type> typeParameters, final Class<?> recordClass) {
+            return type instanceof TypeVariable<?> typeVariable
+                    ? typeParameters.get(getIndexOfTypeParameter(typeVariable.getName(), recordClass))
+                    : type;
+        }
+
+        // Get
+        private static int getIndexOfTypeParameter(final String typeParameterName, final Class<?> recordClass) {
```
Comment (jyemin) [inline bson-record-codec/src/main/org/bson/codecs/record/RecordCodec.java:137]: This looks a bit strange, but the reflection API requires that you match up type parameters by their index in the respective arrays.

**Comment by jyemin** [comment]
Initially I didn't consider the case where the type parameter was used directly as the type of one of the record components.  I wonder if I'm missing anything else.  Please try to think of other test cases.

**Comment by jyemin** [comment]
Unfortunately, this doesn't handle the case of a parameterized record with cyclical type references.  If you attempt to create codecs for this set of records:

```
record TestSelfReferentialHolderRecord(
                                             @BsonId String id,
                                              TestSelfReferentialRecord<String> selfReferentialRecord) {}


public record TestSelfReferentialRecord<T>(
                                        T name,
                                        @Nullable TestSelfReferentialRecord<T> left,
                                        @Nullable TestSelfReferentialRecord<T> right) {}
```

you'll get an infinite recursion resulting in a StackOverflowException:

```
	at org.bson.codecs.record.RecordCodec.<init>
	at org.bson.codecs.record.RecordCodec.parameterize
	at org.bson.codecs.record.RecordCodec$ComponentModel.computeCodec
	at org.bson.codecs.record.RecordCodec$ComponentModel.<init>
	at org.bson.codecs.record.RecordCodec.getComponentModels
	at org.bson.codecs.record.RecordCodec.<init>
```
and so on.  There is nothing to break the cycle.  

To fix this we need to break the cycle.  But I think that can be the subject of another ticket and PR.  This one adds value even if it doesn't handle cycles.

**Review by rozza — CHANGES_REQUESTED** [review]
Add a regression test case for the resolving of types which move.

Add a ticket for the self referential example

File: bson-record-codec/src/main/org/bson/codecs/record/RecordCodec.java (hunk 2 above)
Comment (jyemin) [reply bson-record-codec/src/main/org/bson/codecs/record/RecordCodec.java:137]: Thanks, I like this test.  

Done

**Review by rozza — APPROVED** [review]
One nit - location of the records.

Other than that LGTM

File: bson-record-codec/src/test/unit/org/bson/codecs/record/RecordCodecTest.java [hunk 3]
```diff
@@ -261,6 +262,48 @@ public void testRecordWithNestedParameterizedRecord() {
         assertEquals(testRecord, decoded);
     }
 
+    record TestRecordWithNestedParameterizedRecord(
```
[the remark this answers is not in the document]
Comment (jyemin) [reply bson-record-codec/src/test/unit/org/bson/codecs/record/RecordCodecTest.java:265]: Oh, I was going to ask you about that, and then I forgot.  It seems better, now that I realize it's possible, to just embed the records in line like this.  If you agree, I would just move them all in a separate commit.  But either way, I can move these to where the rest of them are, to be consistent.

[the remark this answers is not in the document]
Comment (rozza) [reply bson-record-codec/src/test/unit/org/bson/codecs/record/RecordCodecTest.java:265]: Happy either way - its nice and accessible that they are inline with the tests so I'm leaning towards this change.

**Review by rozza — APPROVED** [review]
LGTM

File: bson-record-codec/src/test/unit/org/bson/codecs/record/RecordCodecTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply bson-record-codec/src/test/unit/org/bson/codecs/record/RecordCodecTest.java]: ok, i will follow up with that later.

**Review by katcharov — APPROVED** [review]
LGTM

**Pull request merged**